src/dbnode/storage/index/block.go

```diff
@@ -406,54 +396,149 @@ func (b *block) Query(
+	exhaustive, err := b.queryNoLock(ctx, cancellable, query, opts, results)
 	if err != nil {
 		sp.LogFields(opentracinglog.Error(err))
 	}
 
 	return exhaustive, err
 }
 
-func (b *block) queryWithSpan(
+const (
+	queryGroupReadersParallelism = 8
+	queryGroupSize               = 8
+)
+
+type queryGroup struct {
+	readers    []m3ninxindex.Reader
+	exhaustive bool
+	err        error
+}
+
+func (b *block) segmentReadersNoLock() ([]segment.Reader, error) {
+	b.RLock()
+	defer b.RUnlock()
+
+	if b.state == blockStateClosed {
+		return nil, ErrUnableToQueryBlockClosed
+	}
+
+	return b.segmentReadersWithRLock()
+}
+
+func (b *block) queryNoLock(
 	ctx context.Context,
-	cancellable *xresource.CancellableLifetime,
+	cancellable *resource.CancellableLifetime,
 	query Query,
 	opts QueryOptions,
 	results BaseResults,
-	sp opentracing.Span,
-	logFields []opentracinglog.Field,
 ) (bool, error) {
-	b.RLock()
-	defer b.RUnlock()
+	readers, err := b.segmentReadersNoLock()
+	if err != nil {
+		return false, err
+	}
 
-	if b.state == blockStateClosed {
-		return false, ErrUnableToQueryBlockClosed
+	segmentReaders := make([]m3ninxindex.Reader, 0, len(readers))
+	for _, reader := range readers {
+		segmentReaders = append(segmentReaders, reader)
+	}
+
+	if len(segmentReaders) < queryGroupReadersParallelism {
+		// Query no parallelism, but ensure not to overwhlem by limiting
+		// concurrency to the query segments worker pool.
+		b.querySegmentsWorkers.GetToken()
+		exhaustive, err := b.queryReadersNoLock(ctx, cancellable, query,
+			opts, results, segmentReaders)
+		b.querySegmentsWorkers.PutToken()
+		return exhaustive, err
+	}
+
+	var (
+		groupsN = int(math.Ceil(float64(len(readers)) / float64(queryGroupSize)))
+		groups  = make([]queryGroup, groupsN)
+		jobs    = make([]m3ninxindex.Reader, groupsN*queryGroupSize)
+		wg      sync.WaitGroup
+	)
+	// Create query group jobs.
+	for i := 0; i < groupsN; i++ {
+		groupJobs := jobs[:queryGroupSize]
+		jobs = jobs[queryGroupSize:]
+		groups[i] = queryGroup{
+			// Jobs backed by single bulk alloc slice, but start zero length.
+			readers: groupJobs[:0],
+		}
+	}
+	// Allocate jobs to groups, first sort by size.
+	sort.Slice(segmentReaders, func(i, j int) bool {
+		nI, _ := segmentReaders[i].NumDocs()
+		nJ, _ := segmentReaders[j].NumDocs()
+		return nI < nJ
+	})
+	// Now allocate round robin.
+	for i, reader := range segmentReaders {
+		group := i % groupsN
+		groups[group].readers = append(groups[group].readers, reader)
+	}
+
+	// Launch async queries.
+	for i := 1; i < groupsN; i++ {
+		i := i
+		wg.Add(1)
+		b.querySegmentsWorkers.Go(func() {
+			exhaustive, err := b.queryReadersNoLock(ctx, cancellable, query,
+				opts, results, groups[i].readers)
+			groups[i].exhaustive, groups[i].err = exhaustive, err
+			wg.Done()
+		})
 	}
 
-	exec, err := b.newExecutorWithRLockFn()
+	// Save an extra goroutine to execute synchronously on local goroutine.
+	b.querySegmentsWorkers.GetToken()
```

Comment:
Should this take token before the async loop? Otherwise local goroutine could potentially be blocked until we finish some async calls